File: tests/rbac/fixtures/audit.py
"""Audit log assertion helpers (Python side) — spec 102 T018.

Audit storage is owned by audit-service. These helpers query the service API
instead of tailing MongoDB collections.
"""

from __future__ import annotations

import datetime as _dt
import hashlib
import json
import os
import time
import urllib.parse
import urllib.request
from typing import Any

REQUIRED_REASONS = {
    "OK",
    "OK_ROLE_FALLBACK",
    "OK_BOOTSTRAP_ADMIN",
    "DENY_NO_CAPABILITY",
    "DENY_PDP_UNAVAILABLE",
    "DENY_INVALID_TOKEN",
    "DENY_RESOURCE_UNKNOWN",
}


def _audit_service_url() -> str:
    value = os.environ.get("AUDIT_SERVICE_URL") or os.environ.get(
        "E2E_AUDIT_SERVICE_URL", "http://localhost:8010"
    )
    return value.rstrip("/")


def _subject_hash(sub: str) -> str:
    # assisted-by Codex Codex-sonnet-4-6
    salt = os.environ.get("AUDIT_SUBJECT_SALT", "caipe-098-audit")
    digest = hashlib.sha256(f"{salt}:{sub}".encode("utf-8")).hexdigest()
    return f"sha256:{digest}"


def _query_audit_records(params: dict[str, str], limit: int = 50) -> list[dict[str, Any]]:
    query = {
        "limit": str(limit),
        "since": (_dt.datetime.now(tz=_dt.UTC) - _dt.timedelta(days=1)).isoformat(),
        **params,
    }
    url = f"{_audit_service_url()}/v1/audit/events?{urllib.parse.urlencode(query)}"
    with urllib.request.urlopen(url, timeout=3) as response:  # noqa: S310
        payload = json.loads(response.read().decode("utf-8"))
    records = payload.get("records", [])
    return records if isinstance(records, list) else []


def assert_audit_record(
    user_id: str,
    resource: str,
    scope: str,
    allowed: bool,
    reason: str,
    *,
    timeout_s: float = 5.0,
    poll_interval_s: float = 0.1,
) -> dict[str, Any]:
    """Block until a matching audit record appears in audit-service."""
    if reason not in REQUIRED_REASONS:
        raise AssertionError(f"reason {reason!r} not in canonical enum {REQUIRED_REASONS!r}")

    deadline = time.monotonic() + timeout_s
    last_seen: dict[str, Any] | None = None
    action = f"{resource}#{scope}"
    outcome = "allow" if allowed else "deny"
    while time.monotonic() < deadline:
        records = _query_audit_records(
            {
                "subject_hash": _subject_hash(user_id),
                "action": action,
                "outcome": outcome,
                "reason_code": reason,
            }
        )
        last_seen = records[0] if records else last_seen
        for record in records:
            if (
                (record.get("action") or record.get("capability")) == action
                and record.get("outcome") == outcome
                and record.get("reason_code") == reason
            ):
                return record
        time.sleep(poll_interval_s)

    raise AssertionError(
        "no matching audit record within "
        f"{timeout_s}s — looked for "
        f"userId={user_id!r}, resource={resource!r}, scope={scope!r}, "
        f"allowed={allowed!r}, reason={reason!r}; last seen: {last_seen!r}"
    )


def clear_audit_log() -> int:
    """Compatibility no-op; audit-service is append-only for tests."""
    return 0


def latest_audit_record_for(user_id: str) -> dict[str, Any] | None:
    """Return the most-recent audit event for a user, or None."""
    records = _query_audit_records({"subject_hash": _subject_hash(user_id)}, limit=1)
    return records[0] if records else None


def now_iso() -> str:
    """ISO-8601 UTC timestamp matching the audit schema's `ts` format."""
    return _dt.datetime.now(tz=_dt.UTC).isoformat().replace("+00:00", "Z")

File: tests/rbac/fixtures/audit.ts
/**
 * Audit log assertion helpers (TypeScript side) — spec 102 T018.
 *
 * Audit storage is owned by audit-service. These helpers query the service
 * API instead of tailing MongoDB collections.
 */

import { createHash } from "crypto";

export const REQUIRED_REASONS = new Set([
  "OK",
  "OK_ROLE_FALLBACK",
  "OK_BOOTSTRAP_ADMIN",
  "DENY_NO_CAPABILITY",
  "DENY_PDP_UNAVAILABLE",
  "DENY_INVALID_TOKEN",
  "DENY_RESOURCE_UNKNOWN",
] as const);

export type AuditReason =
  | "OK"
  | "OK_ROLE_FALLBACK"
  | "OK_BOOTSTRAP_ADMIN"
  | "DENY_NO_CAPABILITY"
  | "DENY_PDP_UNAVAILABLE"
  | "DENY_INVALID_TOKEN"
  | "DENY_RESOURCE_UNKNOWN";

export interface AuditRecord {
  audit_event_id?: string;
  userId?: string;
  userEmail?: string;
  user_email?: string;
  subject_hash?: string;
  resource?: string;
  scope?: string;
  action?: string;
  capability?: string;
  allowed?: boolean;
  outcome?: "allow" | "deny" | "success" | "error";
  reason?: AuditReason;
  reason_code?: AuditReason;
  source: string;
  service?: string;
  route?: string;
  requestId?: string;
  pdp?: "keycloak" | "local" | "cache" | "openfga";
  ts: Date | string;
}

function auditServiceUrl(): string {
  return (process.env.AUDIT_SERVICE_URL ?? process.env.E2E_AUDIT_SERVICE_URL ?? "http://localhost:8010").replace(
    /\/$/,
    "",
  );
}

function subjectHash(sub: string): string {
  // assisted-by Codex Codex-sonnet-4-6
  const salt = process.env.AUDIT_SUBJECT_SALT ?? "caipe-098-audit";
  return `sha256:${createHash("sha256").update(`${salt}:${sub}`).digest("hex")}`;
}

async function queryAuditRecords(params: Record<string, string>, limit = 50): Promise<AuditRecord[]> {
  const url = new URL(`${auditServiceUrl()}/v1/audit/events`);
  url.searchParams.set("limit", String(limit));
  url.searchParams.set("since", new Date(Date.now() - 24 * 60 * 60 * 1000).toISOString());
  for (const [key, value] of Object.entries(params)) {
    url.searchParams.set(key, value);
  }

  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`audit-service query failed: ${response.status} ${await response.text()}`);
  }
  const body = (await response.json()) as { records?: AuditRecord[] };
  return body.records ?? [];
}

export interface AssertOptions {
  timeoutMs?: number;
  pollIntervalMs?: number;
}

export async function assertAuditRecord(
  userId: string,
  resource: string,
  scope: string,
  allowed: boolean,
  reason: AuditReason,
  opts: AssertOptions = {},
): Promise<AuditRecord> {
  if (!REQUIRED_REASONS.has(reason)) {
    throw new Error(`reason ${reason} not in canonical enum`);
  }
  const timeoutMs = opts.timeoutMs ?? 5000;
  const pollIntervalMs = opts.pollIntervalMs ?? 100;
  const deadline = Date.now() + timeoutMs;
  let lastSeen: AuditRecord | null = null;
  const action = `${resource}#${scope}`;
  const outcome = allowed ? "allow" : "deny";

  while (Date.now() < deadline) {
    const records = await queryAuditRecords({
      subject_hash: subjectHash(userId),
      action,
      outcome,
      reason_code: reason,
    });
    lastSeen = records[0] ?? lastSeen;
    const match = records.find(
      (record) =>
        (record.action ?? record.capability) === action &&
        record.outcome === outcome &&
        record.reason_code === reason,
    );
    if (match) return match;
    await new Promise((resolve) => setTimeout(resolve, pollIntervalMs));
  }

  throw new Error(
    `no matching audit record within ${timeoutMs}ms — looked for ` +
      `userId=${userId}, resource=${resource}, scope=${scope}, ` +
      `allowed=${allowed}, reason=${reason}; last seen=${JSON.stringify(lastSeen)}`,
  );
}

export async function clearAuditLog(): Promise<number> {
  // Audit-service storage is append-only from the test client's perspective.
  return 0;
}

export async function latestAuditRecordFor(userId: string): Promise<AuditRecord | null> {
  const records = await queryAuditRecords({ subject_hash: subjectHash(userId) }, 1);
  return records[0] ?? null;
}
